client.go

```diff
@@ -1,48 +1,273 @@
 	return &S3{
-		cli:        client,
-		BucketName: bucketName,
+		sdkClient:    sdkClient,
+		cryptoClient: cryptoClient,
+		bucketName:   bucketName,
+		region:       region,
 	}
 }
 
-// Get returns an io.ReadCloser instance for the given fully qualified S3 URL.
-// Note that this function will mutate the bucket name of the S3 object if it's different than the existing one.
-func (s3 *S3) Get(rawURL string) (io.ReadCloser, error) {
+// BucketName is a getter for the bucket name used by this S3 client
+func (cli *S3) BucketName() string {
+	return cli.bucketName
+}
+
+// Upload handles the uploading a file to AWS S3, into the bucket configured for this client
+func (cli *S3) Upload(ctx context.Context, req *UploadRequest, payload []byte) error {
+	return cli.UploadWithPsk(ctx, req, payload, nil)
+}
+
+// UploadWithPsk handles the uploading a file to AWS S3, into the bucket configured for this client, using a user-defined psk
+func (cli *S3) UploadWithPsk(ctx context.Context, req *UploadRequest, payload []byte, psk []byte) error {
+
+	listMultiInput := &s3.ListMultipartUploadsInput{
+		Bucket: &cli.bucketName,
+	}
+
+	listMultiOutput, err := cli.sdkClient.ListMultipartUploads(listMultiInput)
+	if err != nil {
+		log.Event(ctx, "error fetching multipart list", log.Error(err))
+		return err
+	}
+
+	var uploadID string
+	for _, upload := range listMultiOutput.Uploads {
+		if *upload.Key == req.UploadKey {
+			uploadID = *upload.UploadId
+		}
+	}
+
+	if len(uploadID) == 0 {
+
+		createMultiInput := &s3.CreateMultipartUploadInput{
+			Bucket:      &cli.bucketName,
+			Key:         &req.UploadKey,
+			ContentType: &req.Type,
+		}
+
+		createMultiOutput, err := cli.sdkClient.CreateMultipartUpload(createMultiInput)
+		if err != nil {
+			log.Event(ctx, "error creating multipart upload", log.Error(err))
+			return err
+		}
+
+		uploadID = *createMultiOutput.UploadId
+	}
+
+	rs := bytes.NewReader(payload)
+
+	n := int64(req.ChunkNumber)
+
+	uploadPartInput := &s3.UploadPartInput{
+		UploadId:   &uploadID,
+		Bucket:     &cli.bucketName,
+		Key:        &req.UploadKey,
+		Body:       rs,
+		PartNumber: &n,
+	}
+
+	if psk != nil {
+		_, err = cli.cryptoClient.UploadPartWithPSK(uploadPartInput, psk)
+		if err != nil {
+			log.Event(ctx, "error uploading with psk", log.Error(err))
+			return err
+		}
+	} else {
+		_, err = cli.sdkClient.UploadPart(uploadPartInput)
+		if err != nil {
+			log.Event(ctx, "error uploading part", log.Error(err))
+			return err
+		}
+	}
+
+	log.Event(ctx, "chunk accepted", log.Data{
+		"chunk_number": req.ChunkNumber,
+		"max_chunks":   req.TotalChunks,
+		"file_name":    req.FileName,
+	})
+
+	input := &s3.ListPartsInput{
+		Key:      &req.UploadKey,
+		Bucket:   &cli.bucketName,
+		UploadId: &uploadID,
+	}
+
+	output, err := cli.sdkClient.ListParts(input)
+	if err != nil {
+		log.Event(ctx, "error listing parts", log.Error(err))
+		return err
+	}
+
+	parts := output.Parts
+	if len(parts) == req.TotalChunks {
```

Comment:
Do we need to do anything if the number of parts is not the same?

Reply:
No, we will only reach this point in the function if the upload-chunk was accepted. If all parts have been uploaded, then we call complete upload. If not all parts have been uploaded yet, we don't complete it yet.